openbas-api/src/main/java/io/openbas/service/ChallengeService.java

```diff
@@ -124,57 +139,58 @@ public ChallengesReader playerChallenges(String exerciseId, User user) {
       String exerciseId, String challengeId, ChallengeTryInput input, User user) {
     ChallengeResult challengeResult = tryChallenge(challengeId, input);
-    if (challengeResult.getResult()) {
+    if (challengeResult.isResult()) {
       // Find and update the expectations linked to the user
       List<InjectExpectation> playerExpectations =
           injectExpectationRepository.findByUserAndExerciseAndChallenge(
               user.getId(), exerciseId, challengeId);
       playerExpectations.forEach(
           playerExpectation -> {
-            playerExpectation.setScore(playerExpectation.getExpectedScore());
-            InjectExpectationResult expectationResult =
-                InjectExpectationResult.builder()
-                    .sourceId("challenge")
-                    .sourceType("challenge")
-                    .sourceName("Challenge validation")
-                    .result(Instant.now().toString())
-                    .date(Instant.now().toString())
-                    .score(playerExpectation.getExpectedScore())
-                    .build();
-            playerExpectation.getResults().add(expectationResult);
-            playerExpectation.setUpdatedAt(Instant.now());
-            injectExpectationRepository.save(playerExpectation);
+            InjectStatus injectStatus = playerExpectation.getInject().getStatus().orElseThrow();
+            ChallengeAttemptId challengeAttemptId =
+                challengeAttemptId(challengeId, injectStatus.getId(), user.getId());
+            ChallengeAttempt challengeAttempt =
+                this.challengeAttemptService
+                    .challengeAttempt(challengeAttemptId)
+                    .orElse(challengeAttempt(challengeAttemptId));
+            double score =
+                challengeAttempt.getAttempt() < playerExpectation.getChallenge().getMaxAttempts()
+                    ? playerExpectation.getExpectedScore()
+                    : 0;
+
+            ExpectationUpdateInput expectationUpdateInput = buildChallengeUpdateInput(score);
+            this.injectExpectationService.updateInjectExpectation(
+                playerExpectation.getId(), expectationUpdateInput);
+
+            // Clean attempt
+            this.challengeAttemptService.deleteChallengeAttempt(challengeAttemptId);
+          });
+    } else {
+      List<InjectExpectation> playerExpectations =
+          injectExpectationRepository.findByUserAndExerciseAndChallenge(
+              user.getId(), exerciseId, challengeId);
+      playerExpectations.forEach(
+          playerExpectation -> {
+            InjectStatus injectStatus = playerExpectation.getInject().getStatus().orElseThrow();
+            ChallengeAttemptId challengeAttemptId =
+                challengeAttemptId(challengeId, injectStatus.getId(), user.getId());
+            ChallengeAttempt challengeAttempt =
+                this.challengeAttemptService
+                    .challengeAttempt(challengeAttemptId)
```

Comment:
question: Could we do this in a one call instead of doing it every player's expectation? If there are a lot, it could be a performance issue as we do a bindById inside

Reply:
I made changes, what do you think ?